feat: support tensor parallel & Data loader

kmehant opened the pull request:
# What does this PR do?

1. Implements `TorchTensorParallelPlugin` to support TP with Pytorch 2.0. This work should be seen along with the PR https://github.com/huggingface/transformers/pull/34194.
2. https://github.com/huggingface/transformers/pull/34184
3. Modifies dataloader to support passing same samples across TP ranks

Please review in conjunction with https://github.com/huggingface/transformers/pull/34194

# Results

See significant improvement in both memory and throughput compared against single gpu training, and FSDP across different settings (checkpointing on/off) and context lengths.

Done on two models
1. ibm-granite/granite-8b-code-base-128k 
4. codellama/CodeLlama-7b-hf

Tables below show the max cuda memory and throughput for various configurations showing the potential of TP contributed in this PR. There is gains in both memory and throughput.

**Note:** Please be aware that the effective TPS for FSDP would be multiplicative of the parallel factor (number of GPUs/devices engaged in distributed training) whereas that is not the case with TP. Therefore, when effective throughput is considered we can find FSDP is better than TP in terms of throughput. However, that **may** be compensated by increasing the batch size utilizing the memory gains etc. 


Model | Method | # of GPUs | Context Length | Batch Size | Grad Checkpointing | Cuda Max Mem (GiB) | Tokens/Sec/GPU
-- | -- | -- | -- | -- | -- | -- | --
ibm-granite/granite-8b-code-base-128k | Single GPU non-distributed | 1 | 8192 | 1 | FALSE | OOM | NA
ibm-granite/granite-8b-code-base-128k | FSDP | 4 | 8192 | 1 | FALSE | OOM | NA
ibm-granite/granite-8b-code-base-128k | TP (This PR) | 4 | 8192 | 1 | FALSE | 52.4 | 7675.4

Model | Method | # of GPUs | Context Length | Batch Size | Grad Checkpointing | Cuda Max Mem (GiB) | Tokens/Sec/GPU
-- | -- | -- | -- | -- | -- | -- | --
ibm-granite/granite-8b-code-base-128k | Single GPU non-distributed | 1 | 8192 | 1 | TRUE | OOM | NA
ibm-granite/granite-8b-code-base-128k | FSDP | 4 | 8192 | 1 | TRUE | 29.975586 | 2256.896
ibm-granite/granite-8b-code-base-128k | TP (This PR) | 4 | 8192 | 1 | TRUE | 26.5 | 5935.5

Model | Method | # of GPUs | Context Length | Batch Size | Grad Checkpointing | Cuda Max Mem (GiB) | Tokens/Sec/GPU
-- | -- | -- | -- | -- | -- | -- | --
ibm-granite/granite-8b-code-base-128k | Single GPU non-distributed | 1 | 16384 | 1 | FALSE | OOM | NA
ibm-granite/granite-8b-code-base-128k | FSDP | 4 | 16384 | 1 | FALSE | OOM | NA
ibm-granite/granite-8b-code-base-128k | TP (This PR) | 4 | 16384 | 1 | FALSE | OOM | NA

Model | Method | # of GPUs | Context Length | Batch Size | Grad Checkpointing | Cuda Max Mem (GiB) | Tokens/Sec/GPU
-- | -- | -- | -- | -- | -- | -- | --
ibm-granite/granite-8b-code-base-128k | Single GPU non-distributed | 1 | 16384 | 1 | TRUE | OOM | NA
ibm-granite/granite-8b-code-base-128k | FSDP | 4 | 16384 | 1 | TRUE | 36.8 | 2084.864
ibm-granite/granite-8b-code-base-128k | TP (This PR) | 4 | 16384 | 1 | TRUE | 33.5 | 5692.5


Model | Method | # of GPUs | Context Length | Batch Size | Grad Checkpointing | Cuda Max Mem (GiB) | Tokens/Sec/GPU
-- | -- | -- | -- | -- | -- | -- | --
codellama/CodeLlama-7b-hf | Single GPU non-distributed | 1 | 8192 | 1 | FALSE | OOM | NA
codellama/CodeLlama-7b-hf | FSDP | 4 | 8192 | 1 | FALSE | 70.7 | 3560
codellama/CodeLlama-7b-hf | TP (This PR) | 4 | 8192 | 1 | FALSE | 42.8 | 9216

Model | Method | # of GPUs | Context Length | Batch Size | Grad Checkpointing | Cuda Max Mem (GiB) | Tokens/Sec/GPU
-- | -- | -- | -- | -- | -- | -- | --
codellama/CodeLlama-7b-hf | Single GPU non-distributed | 1 | 8192 | 1 | TRUE | 75.3 | 2849
codellama/CodeLlama-7b-hf | FSDP | 4 | 8192 | 1 | TRUE | 26.4 | 5957
codellama/CodeLlama-7b-hf | TP (This PR) | 4 | 8192 | 1 | TRUE | 21.4 | 7125

Model | Method | # of GPUs | Context Length | Batch Size | Grad Checkpointing | Cuda Max Mem (GiB) | Tokens/Sec/GPU
-- | -- | -- | -- | -- | -- | -- | --
codellama/CodeLlama-7b-hf | Single GPU non-distributed | 1 | 16384 | 1 | FALSE | OOM | NA
codellama/CodeLlama-7b-hf | FSDP | 4 | 16384 | 1 | FALSE | OOM | NA
codellama/CodeLlama-7b-hf | TP (This PR) | 4 | 16384 | 1 | FALSE | OOM | NA

Model | Method | # of GPUs | Context Length | Batch Size | Grad Checkpointing | Cuda Max Mem (GiB) | Tokens/Sec/GPU
-- | -- | -- | -- | -- | -- | -- | --
codellama/CodeLlama-7b-hf | Single GPU non-distributed | 1 | 16384 | 1 | TRUE | 75.3 | 2599
codellama/CodeLlama-7b-hf | FSDP | 4 | 16384 | 1 | TRUE | 30.1 | 2433
codellama/CodeLlama-7b-hf | TP (This PR) | 4 | 16384 | 1 | TRUE | 26.6 | 6873


Fixes # (issue)
https://github.com/huggingface/transformers/issues/32470

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

I have cycles to bring in more improvements over this PR to bring in Pytorch TP support to HF. Looking forward. Thank you

muellerzr reviewed (approved):
Thanks! This looks great to me. We do still need to update this to work with `accelerate config` however, whcih happens in `commands/config` and `commands/launch`. Would you like to do so?

muellerzr commented:
@kmehant if you rebase from `main` this should fix the failures (tl;dr we had py 3.8 EOL)

kmehant commented:
@muellerzr Appreciate your response. I would like to bring to your notice the below two points.

1. This dataloader written to work for the paradigm (call it paradigm 1) of master process fetching the data needed and distributing them to all the worker processes. The more general paradigm (call it paradigm 2) of all the processes fetching their own data sample in TP case it has to be the same batch across the processes is not covered in this PR.
2. This PR has a soft dependency to apply TP plan over the model since this PR is more like of 2 parts - TP workflow through accelerate plugin + dataloader. 
   1. First part of the PR applies TP parallelism to the model like shown here - https://github.com/huggingface/accelerate/pull/3173/files#diff-2d7515874eaecac2687c7fc1a9c720be53f802bf14b4c3dcebe14ad443d075dcR1467 creating a soft dependency over https://github.com/huggingface/transformers/pull/34194 (Part of this would be superseded by https://github.com/huggingface/transformers/pull/34184 that is carrying a different interface to apply TP plan to the model).
   2. second part is the dataloader

For point (1) I can keep this PR simple and allow only for the paradigm 1 and address the paradigm 2 in another PR.
For point (2) I can remove application of TP part from this PR, keeping this simple and independent. The part removed can be added in a separate PR as point (2)(i) is completed.


WDYT?

BenjaminBossan commented on src/accelerate/data_loader.py:
```diff
@@ -1009,6 +1040,8 @@ def prepare_data_loader(
             "If set to true, the dataloader prepared by the Accelerator will be backed by "
             "[torchdata.StatefulDataLoader](https://github.com/pytorch/data/tree/main/torchdata/stateful_dataloader).
             This requires `torchdata` version 0.8.0 or higher that supports StatefulDataLoader to be installed."
+        torch_device_mesh (``, *optional*, defaults to `None`):
```
The type is missing.

kmehant replied:
Added, thank you!

BenjaminBossan commented on src/accelerate/utils/dataclasses.py:
```diff
@@ -1810,6 +1811,31 @@ def set_mixed_precision(self, mixed_precision, buffer_autocast=False, override=F
+@dataclass
+class TorchTensorParallelPlugin:
+    """
+    This plugin is used to enable tensor parallelism using PyTorch 2.0.
```
I think a reference to further information on TP should be added here. Also:

```suggestion
    This plugin is used to enable tensor parallelism using PyTorch >= 2.0.
```

BenjaminBossan commented on src/accelerate/utils/dataclasses.py:
```diff
@@ -1810,6 +1811,31 @@ def set_mixed_precision(self, mixed_precision, buffer_autocast=False, override=F
+    )
+
+    def __post_init__(self):
+        from torch.distributed.device_mesh import init_device_mesh
```
Should we perform a check on the minimum PyTorch and transformers versions? Not sure if here is the best place or somewhere else, Zach?

BenjaminBossan replied:
I see, good point. Still, torch could be checked, right?

kmehant replied:
Added torch version check thanks

BenjaminBossan reviewed (changes requested):
Thanks for this PR, this looks nice. I have a few smaller comments, please take a look.

Also, please ensure that `make quality` passes.

kmehant commented on src/accelerate/accelerator.py:
```diff
@@ -1457,6 +1463,8 @@ def prepare_model(self, model: torch.nn.Module, device_placement: bool = None, e
                     if self.ddp_handler is not None:
                         self.ddp_handler.register_comm_hook(model)
+            elif self.distributed_type == DistributedType.TP:
+                model.apply_tensor_parallel(self.state.torch_tp_plugin.torch_device_mesh["tp"])
```
Hi @BenjaminBossan 

The `tensor_parallel()` interface will be implemented here - https://github.com/huggingface/transformers/pull/34184/files#diff-6b72b98c4c2dcfc6cc606843917733f5d858374fbc22a735ff483bbc0c1e63eaR5017

I have raised a comment on providing a way to know if tensor_parallel succeeded or not. Once that PR is ready, we can handle it here. WDYT?

BenjaminBossan replied:
Okay, let's see what the final result will be. But we could also check `hasattr(model, "apply_tensor_parallel")` or would that not work?

BenjaminBossan replied:
Ah I see, in that case it is crucial to add a method or attribute to check the support for TP.

kmehant replied:
@BenjaminBossan 

`has_tp_plan` property is added, so updated the code here to fail when the model has no support thank you.

kmehant commented on src/accelerate/data_loader.py:
```diff
@@ -740,7 +742,29 @@ def _fetch_batches(self, iterator):
         batches, batch = None, None
         # On process 0, we gather the batch to dispatch.
         if self.state.process_index == 0:
+            # if a device mesh is provided extract each dimension (tp and dp)
```
Made changes, Thankyou.

kmehant commented:
@muellerzr can I work on this https://github.com/huggingface/accelerate/pull/3173#pullrequestreview-2401793359 in a separate PR?

I have fetched and rebased my PR and addressed all the review comments thank you.

HoangCongDuc commented:
This feature is really useful, thank you @kmehant. I wonder if it is possible to combine tensor parallel with data parallel after this PR, say, TP for same-node parallelism and DP for multi-node parallelism.

kmehant commented:
> This feature is really useful, thank you @kmehant. I wonder if it is possible to combine tensor parallel with data parallel after this PR, say, TP for same-node parallelism and DP for multi-node parallelism.


Hi @HoangCongDuc, support for that is in my TODOs but not covered in this PR, should be coming soon after discussing with HF. Thank you.

kmehant commented on src/accelerate/accelerator.py:
```diff
@@ -1461,6 +1467,10 @@ def prepare_model(self, model: torch.nn.Module, device_placement: bool = None, e
                     if self.ddp_handler is not None:
                         self.ddp_handler.register_comm_hook(model)
+            elif self.distributed_type == DistributedType.TP:
+                if not model.has_tp_plan:
```
@BenjaminBossan 
Yes, it got modified. I have updated this PR again and also that PR to transformers is now merged :)

muellerzr reviewed (commented):
Thanks! Overall the code looks sound, what I'd appreciate however is if we could bring this the last 10% of the way through:

1. Actually implementing this in the CLI and setting the env variable up properly
2. Writing some tests (`src/accelerate/test_utils/scripts/test_tensor_parallel.py` IMO)

kmehant commented:
@muellerzr @BenjaminBossan 

1. Have implemented the CLI part
    1. accelerate launch usage
    2. accelerate config usage
2. Have added run test using existing scripts for TP through CLI

Let me know if I have missed out something. Thank you.

BenjaminBossan reviewed (approved):
Thanks for the updates and sorry for the delays. My points have been addressed, so I'm approving the PR. But it would be great if @muellerzr could re-review it with the latest changes before merging.

kmehant commented:
@BenjaminBossan @muellerzr can we open this PR back and further requesting review and merge. Thank you.

SunMarc commented:
Can you fix the conflits and the CI ? We will merge the PR soon after zach re-review it !

SunMarc commented on src/accelerate/accelerator.py:
```diff
@@ -1471,6 +1492,10 @@ def prepare_model(self, model: torch.nn.Module, device_placement: bool = None, e
                     if self.ddp_handler is not None:
                         self.ddp_handler.register_comm_hook(model)
+            elif self.distributed_type == DistributedType.TP:
+                if not model.supports_tp_plan:
```
model don't necessarily have the `supports_tp_plan` attribute. So make sure to check if the attribute exist also.

kmehant replied:
Thank you for the review, I have add a check for this.

SunMarc commented on src/accelerate/accelerator.py:
```diff
@@ -1471,6 +1492,10 @@ def prepare_model(self, model: torch.nn.Module, device_placement: bool = None, e
                         self.ddp_handler.register_comm_hook(model)
+            elif self.distributed_type == DistributedType.TP:
+                if not model.supports_tp_plan:
+                    raise NotImplementedError("Provided model does not support tensor parallelism")
```
Could you give the user some indication of how they can add the support ? Also make sure to check the version of transformers needed as it was added quite recently.

kmehant replied:
> Could you give the user some indication of how they can add the support ? 
I have extended the message including these details

> Also make sure to check the version of transformers needed as it was added quite recently.
Added version based check as well.

Thank you.

SunMarc commented on src/accelerate/accelerator.py:
```diff
@@ -1471,6 +1492,10 @@ def prepare_model(self, model: torch.nn.Module, device_placement: bool = None, e
                     )
                     if self.ddp_handler is not None:
                         self.ddp_handler.register_comm_hook(model)
+            elif self.distributed_type == DistributedType.TP:
+                if not model.supports_tp_plan:
```
For a first version, it's fine if we rely on transformers integration but it would be a nice follow-up to make it compatible with any pytorch model.

SunMarc reviewed (approved):
Couple of comments. Otherwise LGTM !

SunMarc commented on tests/tp/test_tp.py:
```diff
@@ -0,0 +1,61 @@
+set_seed(42)
+
+
+@require_non_torch_xla
+@require_tp
+@require_multi_device
+@slow
```
you need to add another decorator for transformers lib

kmehant replied:
Have added transformers lib check decorator.

SunMarc commented on tests/tp/test_tp.py:
```diff
@@ -0,0 +1,61 @@
+# limitations under the License.
+
+
+from transformers.trainer_utils import set_seed
+
```
make sure to protect the import, the tests are failing bc of that

kmehant replied:
Have moved the import into the test case itself which only runs when transformers lib is available. Thank you.

SunMarc commented on tests/tp/test_tp.py:
```diff
@@ -0,0 +1,61 @@
+        self.test_tp_size = 2
+        self.model_name_or_path = "TinyLlama/TinyLlama-1.1B-Chat-v1.0"
+        self.batch_size = 1
+        from transformers.trainer_utils import set_seed
+
+        set_seed(42)
+
```
you can use `set_seed` from accelerate 

```suggestion
        from accelerate.utils import set_seed

        set_seed(42)

```

SunMarc commented:
There are some failing tests in the CI, can you check if this is related to this PR @kmehant

kmehant commented:
@SunMarc Have fixed the failing test and tested it locally, requesting CI/CD workflow run again on this PR as a finishing check. Thanks.

muellerzr reviewed (approved):
Nicely done! This looks great to me, thanks so much for all your work! I'll keep an eye on main and will ping you if we find `test_tp` fails at some point
